### ISSUE TITLE
OpenFile() in MpqArchive seems to keep file handle open even when calling Close/Dispose

### ISSUE THREAD

**Issue by NightUndead** [Drake53/War3Net#56]
I have a program where I want to load a bunch of files from the main Warcraft 3 files (specifically War3Patch.mpq from version 1.27), and while I trust that no writing is happening to any files, I decided to copy War3Patch.mpq to a temporary folder and work with it from there, just to make sure ("better safe than sorry" sort of thing).
Obviously, once I'm done working with the copied file, I want to delete it, however, it seems something is left open somewhere after calling OpenFile on an MpqArchive.

This code for example:
```
var archive = MpqArchive.Open(filePath);
archive.Dispose();
```
Gives me no issues and I can immediately delete the copied file.

But this code
```
var archive = MpqArchive.Open(filePath);
var otherStream = archive.OpenFile(@"UI\WorldEditStrings.txt");
otherStream.Close();
archive.Dispose();
```
Leaves me unable to delete the file with an "UnauthorizedAccessException". (Refactoring it into an "using" statement does not change the behavior)
I realise IOException, is the exception we should be expecting if it's a stream not being closed properly, but looking at [File.Delete(String)](https://learn.microsoft.com/en-us/dotnet/api/system.io.file.delete?view=net-8.0) the 4 cases where UnauthorizedAccessException would be thrown does not seem to apply to my situation:
_"The caller does not have the required permission."_
Clearly the application must have permission otherwise I shouldn't be able to delete the file in ANY situation, nor would I expect to be able to create the file in my temporary folder in the first place. (I create the temporary folder at runtime as well, so missing permissions seems very unlikely)
_"The file is an executable file that is in use."_ and _"path is a directory."_
It's not an executable nor is it a directory
(the string I'm providing to File.Delete is @".\TemporaryWorkingFolder\War3Patch.mpq")
_"path specified a read-only file."_
And it can't be a read only file, because again, I can delete the file just fine when I don't call OpenFile. Also checking the file in windows when the exception is thrown shows that the file is not marked read only.

To further cement this theory of mine, I downloaded Process Explorer from Sysinternals and looked for the file handle, and it is indeed still open by my application even though calling Close/Dispose should be closing that handle (or atleast I expect it to).

So I tried recompiling just the Mpq aspect of War3Net where I removed the check on _stream.Close() in
```
public override void Close()
{
    base.Close();
    if (_isStreamOwner)
    {
        _stream.Close();
    }
}
```
And just ALWAYS close it, and while that "fix" allowed me to delete the file, it also leaves me unable to close individual streams on the go until I'm completely done.
I can work with my fix, since it basically just means I only need to call dispose at the end, but I thought I'd open an issue so a potential proper fix could be looked at.

**Comment by Drake53** [comment]
The problem is that opening an `MpqArchive` from a path or FileStream will create a `MemoryMappedFile` as optimization. Then in `MpqStream.GetStreamFromMpqArchive`, it creates either a ViewStream if the memory mapped file exists, or it returns the same shared stream of the MpqArchive (which obviously shouldn't be closed if only an MpqStream is closed). The issue is that currently `_isStreamOwner` is always set to false, when it should be true if the MpqStream's base stream is a ViewStream.

Fix: change line 68 in `MpqStream.cs` to `_isStreamOwner = archive.MemoryMappedFile is not null;`

You can also try using the code on the `vjass` branch where this is already fixed. You can see in `MpqStreamFactory.FromArchive()` that `leaveOpen` is set to false for the ViewStream, and true otherwise: https://github.com/Drake53/War3Net/blob/vjass/src/War3Net.IO.Mpq/MpqStreamFactory.cs#L76

**Comment by NightUndead** [comment]
Thanks for the quick response!

So the constructor that line 68 is in, doesn't have an ```archive``` variable available, rather the constructor on line 54 does.
Putting the line ```_isStreamOwner = archive.MemoryMappedFile is not null;``` in that constructor does indeed fix it.
However you mentioned that if the base stream is a ViewStream it should be set to true, so I went a bit further and rather than setting ```_isStreamOwner``` twice, I found that I could also change line 68 to ```_isStreamOwner = baseStream is MemoryMappedViewStream;``` to get a similar fix.
(A similar check is done a few lines below with ```_streamOffset = baseStream is MemoryMappedViewStream ? 0 : _filePosition;```, which is where I got the idea from)

With that change I can run my code with no file access issues.
